2-advanced-topics/data-pipelines/intro-data-pipelines.ipynb

```diff
@@ -0,0 +1,1044 @@
+    "#### Gold\n",
+    "\n",
+    "In the gold stage of the pipeline (gold.py in the project folder), we finalize the dataset for analysis. The focus here is on filtering, selecting key columns, and ensuring the data is structured for consumption.\n",
+    "\n",
+    "This stage refines the data into a ready-to-use form. We retain columns deemed necessary for the analysis, as the remaining columns are available in the silver table if required later."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 0,
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "cellMetadata": {},
+     "inputWidgets": {},
+     "nuid": "3f03058a-35da-43b5-a8c6-c9bd0ac209c8",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "outputs": [],
+   "source": [
+    "gold_df = (spark.read.table(f\"{database_name}.kenya_silver\")\n",
+    "    .filter(col('year') != 2015) # Exclude 2015 data\n",
+    "    .withColumn('country_name', lit('Kenya'))\n",
+    "    .select('country_name',\n",
+    "            'year',\n",
+    "            col('Initial_Budget_Printed_Estimate').alias('approved').cast(DoubleType()),\n",
+    "            col('Final_Budget_Approved_Estimate').alias('revised'),\n",
+    "            col('`Final_Expenditure_Total_Payment_Comm.`').alias('executed'),\n",
+    "            'admin0',\n",
+    "            'admin1',\n",
+    "            'admin2',\n",
+    "            'geo1',\n",
+    "            'is_foreign',\n",
+    "            'func_sub'\n",
+    "           )\n",
+    ")\n",
+    "\n",
+    "# Save to gold table\n",
+    "gold_df.write.format(\"delta\").mode(\"overwrite\").saveAsTable(f\"{database_name}.kenya_gold\")\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "cellMetadata": {},
+     "inputWidgets": {},
+     "nuid": "2b285fd1-0002-4ed8-8736-26165242f9c9",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "source": [
+    "#### Aggregation and additional columns\n",
+    "\n",
+    "In the above steps we cleaned the original source BOOST Kenya data through multiple stages. In addition to this, we might need external datasources to append to this data. In our example, we want to information about percapita spending.\n",
+    "\n",
+    "In this case, we want to get the subnational population from an external source (an API) and then aggregate our kenya_gold data to get a table containing the spending by *geo1* region and year. Then we will append a new column called per_capita_spending to this aggregated table. We finally save this table so that these descriptive figures are readily available. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "cellMetadata": {},
+     "inputWidgets": {},
+     "nuid": "4782332f-c640-49e3-a894-20271a1b6815",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "source": [
+    "##### Subnational Population\n",
```

Comment:
I'd suggest moving this example up to be right before the "The Core of the Data Pipeline.." section. Not only this is a good stand-alone example of a simple data pipeline covering ingestion, processing, and writing, it also provides a quick recap of basic data processing using pandas. The only thing new to someone already know pandas is the writing to the data lake in delta format – it would be a good opportunity to introduce the concept of Spark and delta. Then you can move onto the next section on the basic components / building blocks of a data pipeline and referring back to this simple example and add the orchestration with scheduling demo to show how this can automate updating and saving the subnational population dataset, which can be reused. Then the latter BOOST example can focus on illustrating the medallion architecture and showcasing the re-use/merging with the subnational dataset. This way, we build the concepts on top of each other and gradually introduce them to the participants. Thoughts?

Reply:
It also breaks the instruction/concept only wall of texts up front.